src/file.c

```diff
@@ -531,18 +536,31 @@ size_t RETRACE_IMPLEMENTATION(fwrite)(const void *ptr, size_t size, size_t nmemb
 	event_info.return_value_type = PARAMETER_TYPE_INT;
 	event_info.return_value = &r;
 	event_info.logging_level = RTR_LOG_LEVEL_NOR;
-	retrace_log_and_redirect_before(&event_info);
 
-	if (rtr_str_inject(STRINJECT_FUNC_FWRITE, (void *) ptr, size * nmemb)) {
+	if (size == 1 && rtr_str_inject(STRINJECT_FUNC_FWRITE, ptr, size * nmemb, &inject_buffer, &inject_len)) {
 		event_info.extra_info = "[redirected]";
 		event_info.event_flags = EVENT_FLAGS_PRINT_RAND_SEED | EVENT_FLAGS_PRINT_BACKTRACE;
 		event_info.logging_level |= RTR_LOG_LEVEL_FUZZ;
+
+		parameter_values[2] = &inject_buffer;
+		parameter_values[4] = &inject_len;
+
+		enable_inject = 1;
 	}
 
-	r = real_fwrite(ptr, size, nmemb, stream);
-	if (r < size * nmemb)
+	retrace_log_and_redirect_before(&event_info);
+
+	r = real_fwrite(enable_inject ? inject_buffer : ptr,
+			size,
+			enable_inject ? inject_len : nmemb,
+			stream);
+
+	if (r < (inject_len > size * nmemb ? size * nmemb : size * inject_len))
```

Comment:
Isn't `inject_len` undefined when `size != 1`? Also, `fwrite returns number of items written.
```
if (r < (enable_inject ? inject_len : nmenb))
```